indev/evdev.c

```diff
@@ -170,42 +178,55 @@ void evdev_read(lv_indev_drv_t * drv, lv_indev_data_t * data)
                 else if(in.value == 1)
                     evdev_button = LV_INDEV_STATE_PR;
             } else if(drv->type == LV_INDEV_TYPE_KEYPAD) {
-		data->state = (in.value) ? LV_INDEV_STATE_PR : LV_INDEV_STATE_REL;
-		switch(in.code) {
-			case KEY_BACKSPACE:
-				data->key = LV_KEY_BACKSPACE;
-				break;
-			case KEY_ENTER:
-				data->key = LV_KEY_ENTER;
-				break;
-			case KEY_UP:
-				data->key = LV_KEY_UP;
-				break;
-			case KEY_LEFT:
-				data->key = LV_KEY_PREV;
-				break;
-			case KEY_RIGHT:
-				data->key = LV_KEY_NEXT;
-				break;
-			case KEY_DOWN:
-				data->key = LV_KEY_DOWN;
-				break;
-			default:
-				data->key = 0;
-				break;
-		}
-		evdev_key_val = data->key;
-		evdev_button = data->state;
-		return ;
-	    }
+#if USE_XKB
+                data->key = xkb_process_key(in.code, in.value != 0);
+#else
+                switch(in.code) {
+                    case KEY_BACKSPACE:
+                        data->key = LV_KEY_BACKSPACE;
+                        break;
+                    case KEY_ENTER:
+                        data->key = LV_KEY_ENTER;
+                        break;
+                    case KEY_PREVIOUS:
```

Comment:
Which keys are `KEY_PREVIOUS` and `KEY_NEXT` are mapped to `-` and `+`, right?

For `LV_KEY_NEXT` `tab` would be more conventional to move the focus.

Reply:
Hm, I'm actually not sure what keys trigger `KEY_PREVIOUS` and `KEY_NEXT`. At least none of the keys on my keyboards seem to. I just found them in the [Linux event code header](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h#L479) and thought they'd be good to map to `LV_KEY_PREV` and `LV_KEY_NEXT`, too. Googling around I found a [forum thread](https://www.vdr-portal.de/forum/index.php?thread/126665-tbs6982-codes-umbiegen-key-previous-key-back/) where somebody receives these keys from an IR remote.

Maybe we could map all of `KEY_RIGHT` / `KEY_NEXT` / `KEY_TAB` to `LV_KEY_NEXT` / `KEY_LEFT` and `KEY_PREVIOUS` to `LV_KEY_PREV`?

Reply:
Not that, `LV_KEY_NEXT` and `LV_KEY_LEFT` are different.
- `LV_KEY_NEXT`: focus on the next object
- `LV_KEY_LEFT`: like left arrow, move the cursor on a text area, adjust slider, etc. 


So I suggest binding 
- `KEY_LEFT/RIGHT/UP/DOWN` to `LV_KEY_LEFT/RIGHT/UP/DOWN` 
- `KEY_TAB` to `LV_KEY_NEXT`
- Nothing to `LV_KEY_PREV` (Shift+Tab could be good but I don't if how difficult it is to handle)

Reply:
Ah, got it. Ok, I pushed a change to:

- map `KEY_TAB` to `LV_KEY_NEXT` _outside_ the XKB path
- map XKB's tab and "left tab" (sent on Shift + Tab) key symbols to `LV_KEY_NEXT` and `LV_KEY_PREV`, respectively

I did not add handling for Shift + Tab _outside_ the XKB path because there's currently no handling of Shift at all there. It wouldn't be hard but if somebody needs Shift + Tab (or capital letters), they're probably better off with `#define USE_XKB 1` anyway.

While at it, I noticed that XKB has separate key symbols for keys pressed on the key/number pad. E.g. there is `XKB_KEY_Return` and also `XKB_KEY_KP_Enter`. I simply added these to the existing cases.

I also left the mapping of `XKB_KEY_Prior` and `XKB_KEY_Next` to `LV_KEY_PREV` and `LV_KEY_NEXT` in. If somebody really has an input device that sends these keys, I suppose that would be the most reasonable mapping.

Let me know if this is ok and I'm happy to adapt it if needed.